R/chat_app.R

```diff
@@ -0,0 +1,216 @@
+#' @export
+chat_app <- function(client, ...) {
+  check_ellmer_chat(client)
+
+  ui <- bslib::page_fillable(
+    chat_mod_ui("chat", client = client, height = "100%"),
+    shiny::actionButton(
+      "close_btn",
+      label = "",
+      class = "btn-close",
+      style = "position: fixed; top: 6px; right: 6px;"
+    )
+  )
+
+  server <- function(input, output, session) {
+    chat_mod_server("chat", client)
+
+    shiny::observeEvent(input$close_btn, {
+      shiny::stopApp()
+    })
+  }
+
+  shiny::shinyApp(ui, server, ...)
+}
+
+check_ellmer_chat <- function(client) {
+  if (!inherits(client, "Chat")) {
+    abort("`client` must be an `ellmer::Chat` object.")
+  }
+}
+
+#' @describeIn chat_app A simple chat app module UI.
+#' @param id The chat module ID.
+#' @param messages Initial messages shown in the chat, used when `client` is not
+#'   provided or when the chat `client` doesn't already contain turns. Passed to
+#'   `messages` in [chat_ui()].
+#' @export
+chat_mod_ui <- function(id, ..., client = NULL, messages = NULL) {
+  if (!is.null(client)) {
+    check_ellmer_chat(client)
+
+    client_msgs <- map(client$get_turns(), function(turn) {
+      content <- ellmer::contents_markdown(turn)
+      if (is.null(content) || identical(content, "")) {
+        return(NULL)
+      }
+      list(role = turn@role, content = content)
+    })
+    client_msgs <- compact(client_msgs)
+
+    if (length(client_msgs)) {
+      if (!is.null(messages)) {
+        warn(
+          "`client` was provided and has initial messages, `messages` will be ignored."
+        )
+      }
+      messages <- client_msgs
+    }
+  }
+
+  shinychat::chat_ui(
+    shiny::NS(id, "chat"),
+    messages = messages,
+    ...
+  )
+}
+
+#' @describeIn chat_app A simple chat app module server.
+#' @export
+chat_mod_server <- function(id, client) {
+  check_ellmer_chat(client)
+
+  append_stream_task <- shiny::ExtendedTask$new(
+    function(client, ui_id, user_input) {
+      promises::promise(function(resolve, reject) {
+        tryCatch(
+          resolve(client$stream_async(user_input)),
+          error = function(err) {
+            reject(err)
+          }
+        )
+      })$then(function(stream) {
+        chat_append(ui_id, stream)
+      })
+    }
+  )
+
+  shiny::moduleServer(id, function(input, output, session) {
+    shiny::observeEvent(input$chat_user_input, {
+      append_stream_task$invoke(
+        client,
+        session$ns("chat"),
+        input$chat_user_input
+      )
+    })
+
+    shiny::observe({
+      if (append_stream_task$status() == "error") {
+        tryCatch(
+          append_stream_task$result(),
+          error = notify_error(session$ns("chat"), session)
+        )
+      }
+    })
+
+    shiny::reactive({
+      if (append_stream_task$status() == "success") {
+        client$last_turn()
+      }
+    })
+  })
+}
+
+notify_error <- function(id, session = shiny::getDefaultReactiveDomain()) {
```

Comment:
Please emit the error/stack to the console as well. I've been frustrated in the past by my chatbot erroring and there being nothing actionable in the console.